### PR TITLE
ARROW-2328: [C++] Writing a slice with feather ignores the offset.

### PR DESCRIPTION
**Pull request by Adriandorr**
Writing a slice from row n of length m of an array to feather would write the first m rows, instead of the rows starting at n. 
Fixed and unit tests added.

### CODE REVIEW

**Review by xhochy — CHANGES_REQUESTED** [review]
You need to run `make format` for some code style fixes to get the build passing.

File: cpp/src/arrow/ipc/feather.cc [hunk 1]
```diff
@@ -603,7 +602,12 @@ class TableWriter::TableWriterImpl : public ArrayVisitor {
       }
 
       if (prim_values.values()) {
-        values_buffer = prim_values.values()->data();
+        if (prim_values.offset() != 0 && (fw_type.bit_width() % 8 != 0)) {
+          return arrow::Status::Invalid(
```
[the remark this answers is not in the document]
Comment (Adriandorr) [reply cpp/src/arrow/ipc/feather.cc]: Thanks for your comments. I've changed the status return code. But I'm having two problems:
* when I run 'make format' it only changes a whole bunch of files I didn't change. Was master broken when I forked?
* SliceStringsRoundTrip fails and I'm struggling to understand which of the various offsets needs to be adjusted. I can probably figure that out tomorrow.
Thanks.

**Comment by Adriandorr** [comment]
On further investigation, there is a second bug, in that the null bitmap isn't aligned with the offset either. Hence the SliceStringsRoundTrip test fails.
Looking at the TableWriter code, to fix this will require creating a new null bitmap, which will require a MemoryPool and hence a change in the api. :(. 
Without the ability to have nulls, writing sliced tables to feather seems to be mostly broken. I could add the pool as an optional argument, or what would be a preferred approach? I'll have a think tomorrow.

File: cpp/src/arrow/ipc/feather.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (Adriandorr) [reply cpp/src/arrow/ipc/feather.cc]: I've now made these changes and more.

File: cpp/src/arrow/ipc/feather.cc [hunk 2]
```diff
@@ -558,8 +577,18 @@ class TableWriter::TableWriterImpl : public ArrayVisitor {
...
+        if (values.offset() > 0) {
+          if (!pool_) {
+            return Status::Invalid(
+                "Require memory pol to write array slice (i.e. offset > 0) with nulls.");
+          }
+          RETURN_NOT_OK(GetTruncatedBitmap(values.offset(), values.length(), null_bitmap,
+                                           pool_, &null_bitmap));
+        }
+
+        RETURN_NOT_OK(WritePadded(stream_.get(), null_bitmap->data(), null_bitmap_size,
```
Comment (xhochy) [inline cpp/src/arrow/ipc/feather.cc]: Wouldn't it be better to extend `WritePadded` or write a second version of it that can handle bit-offsets? Using that way we could avoid the memory allocation for `GetTruncatedBitmap` above.

Comment (Adriandorr) [reply cpp/src/arrow/ipc/feather.cc]: I agree that would be better if it is possible. Same goes for the arrow writing code of course. I can give it a try tomorrow.

Comment (Adriandorr) [reply cpp/src/arrow/ipc/feather.cc]: I've pushed a change to do as you suggested and added a test.

**Comment by wesm** [comment]
Would like to review this more closely, will look when I can

**Comment by Adriandorr** [comment]
I seem to have messed up this branch/pull request by merging the latest changes from arrow/master. Do I need to create a new branch? Any advice appreciated. thanks.

**Comment by Adriandorr** [comment]
I've redone the changes in a new branch and pull request (https://github.com/apache/arrow/pull/1784). Hopefully that is a little cleaner.

**Pull request closed without merging**